### ISSUE TITLE
API for modkit extract read_calls_path output?

### ISSUE THREAD

**Issue by billytcl** [nanoporetech/modkit#194]
Is there an API to get the output of modkit extract's read_calls_path (eg, mod calls after thresholding) to get per read tables in an iterator format?

The reason why I want to do this is because I'm working with some huge datasets, and parsing on a per-read basis is really prohibitively slow (also uses a crazy amount of disk space). It would be way easier to parse it on the fly in Python by iterating per read.

Considering that it writes on a per read basis anyway, do you have a code example where I can capture the output and do all the parsing live instead of writing it to disk? I'm imaging per read to output a table, then doing something to it, then iterating to the next read.

**Comment by ArtRand** [comment]
Hello @billytcl, 

I completely understand. Give me a few and I'll make you a build that can stream the read-calls to stdout similarly to how the "raw extract" can currently. I'm also experimenting with different output formats in an effort to speed up `modkit extract`. Depending on what you're doing, maybe I can add it as a filter on the output.

**Comment by billytcl** [comment]
Thanks! It would be useful if I can get them already thresholded!

**Comment by ArtRand** [comment]
@billytcl by "already thresholded" you mean just emit the calls where the `fail` coluimn is `false`?

**Comment by billytcl** [comment]
Yes that’s right!

**Comment by ArtRand** [comment]
Hello @billytcl,

Please find attached a build that allows the following command that will stream the "read calls" table to stdout:

```bash
$ modkit extract ${modbam} null --read-calls stdout [--pass] [--no-header]
```

Perhaps self-explanatory but `--pass` will restrict the read calls table to only "passing" (i.e. thresholded) calls and `--no-header` will omit the first header line (sometimes this makes parsing easier).

These changes will make it into the release later this week as well.

**Comment by billytcl** [comment]
Thank you! Do you have any ideas on how to effectively parse it
programmatically as an iterator in Python? Perhaps I can read line by line
on stdin and keep track of the read name? Or do you have a better idea?

Billy

**Comment by ArtRand** [comment]
@billytcl 

That's basically what I do, aggregate the lines together until the read_id changes. If I'm doing an expensive computation on the batch of lines, I send that batch on a channel to another thread pool to process in parallel, then send the results to an aggregator. Feel free to email me.

**Comment by ArtRand** [comment]
@billytcl this change has been released in [v0.3.1rc1](https://github.com/nanoporetech/modkit/releases/tag/v0.3.1rc1)
